TS-4717:  Http2 stack explosion.

shinrich opened the pull request:
Added a common state_process_frame_read method to loop over reading frames while there is data available.  The original state_start_frame_read and state_complete_frame_read call into state_process_frame_read so the event handling cases still work.  

Have been running a version on of this code on two of our production boxes for a day.  We haven't had a load surge event, so I doubt we have seen a case that would have caused the stack explosion.  But the performance and error stats seem similar to their peers, so I don't think I have messed up the normal operating case.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/509/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/406/ for details.

maskit commented:
There is something wrong with it. I just applied this patch on master and sent a request with nghttp, and then it aborted.
http://pastebin.com/raw/SYZEckUf

According to the log, the first frame TS received was DATA frame, but it should be SETTINGS frame. I guess the frame header isn't initialized / parsed correctly because the value of frame type of DATA frames is 0.

maskit commented:
I was able to get a HTTP response if I reversed a condition `(inside_frame)` to `(!inside_frame)`, though it still aborts.

A good news is that now I don't see stack explosion with a way I found.

shinrich commented:
Thanks @maskit.  I fixed the sense of inside_frame.  Had made a name change and flipped the sense of the bool.  Thought I had fixed up the patch, after testing on my build, but missed it.  Pushed a new commit with that fixed.

I've been running with only the ink_release asserts, so I didn't see the !closed assert go off in UnixNetVConnection.cc.  I've changed them to ink_release_asserts and compiled with enable-debug, but I'm not able to get the !closed asserts to trigger.  Could you share the command you are using for your testing?

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/510/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/407/ for details.

maskit commented:
I'm just sending a request with nghttp.
```
nghttp -v https://localhost:8443
```

It still happens.
http://pastebin.com/raw/mpsn6EAL

On the client side (nghttp), I don't see any error and response data seems OK.

shinrich commented:
Thanks @maskit.  Looking at your debug messages and comparing it to my own, I realized that I was not running completely up to date with master.   Once got the right version of the code, I saw the same assert that you did.

The problem was that the new loop was always reenabling at the end even if the client side connection had shutdown (by sending a goaway frame).  The reenable called setenable which triggered the !closed assert.  I added a is_client_closed method to Http2ClientSession and used it to protect the call to reenable.  With that change, the asserts don't occur.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/511/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/408/ for details.

masaori335 commented:
I'm seeing crash with h2spec. I don't see this with latest master.

h2spec : https://paste.apache.org/SU02
ATS(w/ --enable-debug) : https://paste.apache.org/mPdu

masaori335 commented:
Sorry, my local master was old. The crash with h2spec(4.3) is not related this Pull-Request.
Filed this issue as [TS-4726](https://issues.apache.org/jira/browse/TS-4726).

maskit commented:
Looks good. 👍 
Now I don't see any assert nor stack explosion except the one @masaori335  reported.

jpeach commented on proxy/http2/Http2ClientSession.h:
```diff
@@ -243,6 +243,12 @@ class Http2ClientSession : public ProxyClientSession
   }
 
+  bool
+  is_client_closed() const
```
Implement this in "ProxyClientSession" by calling "get_netvc()"?

shinrich replied:
Sure, makes sense to push up a level.  Fixing now.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/513/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/410/ for details.
